### PR TITLE
feat: extend SDK with content source maps support [TOL-2043]

### PR DESCRIPTION
**Pull request by 2wce**
## Summary

Content source maps will change the sys object in the entity for the response and this has been implemented as follows

- [x] When creating the Client for the SDK, we have added an option to include alpha features where csm can be enabled/disabled at client level. 
- [x]  We have added a wrapper function which will act like middleware of sorts & check if the alpha feature is enabled before adding the includeContentSourceMaps param to the query to enable CSM
- [x] We have added tests to ensure that the alpha feature can be correctly enabled
- [x] we have also added tests to show that adding Content Source Maps won't break any existing logic for fetching assets & entries including the current supported chain modifers 

## Motivation and Context

This has been added as part of this [CEP](https://contentful.atlassian.net/wiki/spaces/ENG/pages/4538695717/CEP-0195+Content+Source+Maps+in+the+REST+API+CPA) and this [ticket](https://contentful.atlassian.net/browse/TOL-2043?atlOrigin=eyJpIjoiNDk3NWQ4ZTA4NTNkNDI1ZDhlZTU3NGJlYjJkOTI1MWQiLCJwIjoiaiJ9) in order to let users fetch Content Source Maps with the REST API

### CODE REVIEW

File: lib/contentful.ts [hunk 1]
```diff
@@ -109,6 +109,13 @@ export interface CreateClientParams {
    * Interceptor called on every response. Takes Axios response object as an arg.
    */
   responseLogger?: (response: AxiosResponse<any> | Error) => unknown
+
+  /**
+   * Enable alpha features.
+   */
+  alphaFeatures?: {
+    withContentSourceMaps?: boolean
```
Comment (z0al) [inline lib/contentful.ts:117]: It would be helpful If you add a JS Doc here explaining briefly what is this about for customers. It will show up to the developer as a help text in the editor completion menu.

File: lib/create-contentful-api.ts [hunk 2]
```diff
@@ -96,6 +97,15 @@ export default function createContentfulApi<OptionType extends ChainOptions>(
     return baseUrl
   }
 
+  const alphaFeatures = (http.httpClientParams as any as CreateClientParams)?.alphaFeatures
+
+  function maybeEnableSourceMaps(query: Record<string, any> = {}): Record<string, any> {
+    if (alphaFeatures?.withContentSourceMaps) {
+      query.includeContentSourceMaps = true
```
Comment (z0al) [inline lib/create-contentful-api.ts:104]: We should validate against enabling Content Source Maps for CDA and report a user friendly error.

File: test/integration/getEntries.test.ts [hunk 3]
```diff
@@ -410,7 +458,21 @@ function assertLocalizedEntriesResponse(response) {
...
+
   expect(response.items[0].fields.bestFriend['en-US'].fields).toBeDefined()
   expect(response.items[0].fields.bestFriend['en-US'].sys.type).toBe('Entry')
   expect(response.items[0].metadata).toEqual({ tags: [] })
 }
+
+function assertCSMEntriesResponse(response) {
+  expect(response.includes).toBeDefined()
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  expect(response.includes!.Asset).toBeDefined()
```
Comment (z0al) [inline test/integration/getEntries.test.ts:470]: Nit: Are these non-null assertions really needed? Why not use `response.includes?.Asset` ?

File: test/integration/getEntries.test.ts [hunk 4]
```diff
@@ -410,7 +458,21 @@ function assertLocalizedEntriesResponse(response) {
...
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  expect(Object.keys(response.includes!.Asset!).length).toBeGreaterThan(0)
+
+  expect(response.items[0].fields.bestFriend.fields).toBeDefined()
```
Comment (z0al) [inline test/integration/getEntries.test.ts:474]: Nit: can we set `response.items[0]` to a variable and use that?

File: test/unit/contentful.test.ts [hunk 5]
```diff
@@ -116,4 +116,16 @@ describe('contentful', () => {
       'http://some-base-url.com/environments/stage',
     )
   })
+
+  test('Initializes API with alpha features', () => {
```
Comment (z0al) [inline test/unit/contentful.test.ts:120]: Not sure how useful this test is atm. 

What I'm missing so far in other tests is that none of explicitly checks if we are passing the correct query param. 

That's more useful than just checking if we are passing the `alphaFeatures` option down the chain.

**Review by z0al — COMMENTED** [review]
Looks great overall. Just a few pieces to iron out.

File: lib/create-contentful-api.ts (hunk 2 above)
Comment (aodhagan-cf) [reply lib/create-contentful-api.ts:109]: could the `alphaFeatures?.withContentSourceMaps` part be done in the `checkIncludeContentSourceMapsParamIsAllowed` function? Then you could simply check isAllowed

nit: `areAllowed`

Comment (2wce) [reply lib/create-contentful-api.ts:109]: sounds good 👌🏽i had picked `isAllowed` since we are checking if the param is allowed not necessarily if content source maps are allowed but it's fine either way I'll push the change

Comment (aodhagan-cf) [reply lib/create-contentful-api.ts:109]: Ah right makes sense then sorry, yeah also happy with either

**Review by jjolton-contentful — APPROVED** [review]
